tpm2_pytss/utils.py

```diff
@@ -226,3 +237,168 @@ def unwrap(
         )
 
     return s
+
+
+class NoSuchIndex(Exception):
+    """NV index is not defined exception
+
+    Args:
+        index (int): The NV index requested
+    """
+
+    def __init__(self, index):
+        self.index = index
+
+    def __str__(self):
+        return f"NV index 0x{index:08x} does not exist"
+
+
+class nv_read_ek:
+    """NV read callback to be used with create_ek_template
+
+    Args:
+        ectx (ESAPI): The ESAPI context for reading from NV areas
+        auth_handle (ESYS_TR): Handle indicating the source of the authorization. Defaults to the index being read.
+        session1 (ESYS_TR): A session for securing the TPM command (optional). Defaults to ESYS_TR.PASSWORD.
+        session2 (ESYS_TR): A session for securing the TPM command (optional). Defaults to ESYS_TR.NONE.
+        session3 (ESYS_TR): A session for securing the TPM command (optional). Defaults to ESYS_TR.NONE.
+    """
+
+    def __init__(
+        self,
+        ectx: ESAPI,
+        auth_handle: ESYS_TR = None,
+        session1: ESYS_TR = ESYS_TR.PASSWORD,
+        session2: ESYS_TR = ESYS_TR.NONE,
+        session3: ESYS_TR = ESYS_TR.NONE,
+    ):
+        self._ectx = ectx
+        self._auth_handle = auth_handle
+        self._session1 = session1
+        self._session2 = session2
+        self._session3 = session3
+        self._buffer_max = 512
+        self._indices = []
+
+        more = True
+        while more:
+            more, data = self._ectx.get_capability(
+                TPM2_CAP.TPM_PROPERTIES,
+                TPM2_PT.FIXED,
+                4096,
+                session1=session2,
+                session2=session3,
+            )
+            props = data.data.tpmProperties
+            for p in props:
+                if p.property == TPM2_PT_NV.BUFFER_MAX:
+                    self._buffer_max = p.value
+                    more = False
+                    break
+
+        more = True
+        while more:
+            more, data = self._ectx.get_capability(
+                TPM2_CAP.HANDLES,
+                TPM2_HC.HR_NV_INDEX,
+                4096,
+                session1=session2,
+                session2=session3,
+            )
+            self._indices += list(data.data.handles)
+
+    def __call__(self, index: Union[int, TPM2_RH]) -> bytes:
+        if index not in self._indices:
+            raise NoSuchIndex(index)
```

Comment:
What if the index was defined between the list population and now. Perhaps it's better to just try and read the NV index and look at the error code to see if it's not defined. Which also reduces the trips we need to make to the TPM. Which then, do we even need this exception class, we could just use TSS2 Exceptions that already exist.

Reply:
I view the chance as low for this use case as new template/certificate areas shouldn't be created during runtime
The problem with letting the TPM throw an error is that it would break audit sessions (and possibly encrypted sessions) for the scenarios where create_ek_template looks for an index but shouldn't fail if there isn't one
We could have a constant of an TSS2_Exception instead, but something explicit and clear is needed, for example someone writes their own callback using FAPI instead which wouldn't throw the same exception

Reply:
> I view the chance as low for this use case as new template/certificate areas shouldn't be created during runtime

It's an unlikely race condition, but it's a TOCTOU issue.

> The problem with letting the TPM throw an error is that it would break audit sessions (and possibly encrypted sessions) for 

I'm not following, what exactly breaks? Does a failing command invalidate the session?

> the scenarios where create_ek_template looks for an index but shouldn't fail if there isn't one We could have a constant of an TSS2_Exception instead, but something explicit and clear is needed, for example someone writes their own callback using FAPI instead which wouldn't throw the same exception

True, or if they use their own custom TSS.

Reply:
> > I view the chance as low for this use case as new template/certificate areas shouldn't be created during runtime
> 
> It's an unlikely race condition, but it's a TOCTOU issue.
> 
> > The problem with letting the TPM throw an error is that it would break audit sessions (and possibly encrypted sessions) for
> 
> I'm not following, what exactly breaks? Does a failing command invalidate the session?
> 
I reread the specification and I was wrong, so that can be dropped, I'll rework the code to not check the index before trying to read and raise a NoSuchIndex exception if the TSS2_Exception matches the correct error, otherwise reraise the TSS2_Exception